File: DSA/PranyaGupta_500126032/Day27Question1.py
class node:
    def __init__(self, data):
        self.next = None
        self.prev = None
        self.data = data
class LinkedList:
    def __init__(self):
        self.head = None
    def add(self, data):
        new_node = node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
        new_node.prev = current
    def printlist(self, head):
        current = head
        while current:
            print(current.data, end=" ")
            current = current.next   
    def deletenode(self, head, pos):
        if pos == 1:
            self.head = self.head.next
            self.head.prev = None
            self.printlist(self.head)
        i = 1
        current = self.head
        prevn = None
        while  i<pos and current.next:
            prevn = current
            current = current.next
            i+=1
        if not current.next:
            return    
        current.prev = None
        prevn.next = current.next
        current.next.prev = prevn
        self.printlist(head)
       


def main():
    user_input = input("Enter elements of the linked list (with space in between): ")
    arr = list(map(int, user_input.strip().split()))
    pos = int(input("Enter pos: "))
    l = LinkedList()
    for num in arr:
        l.add(num)
    l.deletenode(l.head, pos)
if __name__ == "__main__":
    main()

File: DSA/SAMAGRA_590014891/Day_27_Question_1.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* prev;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

struct Node* insertEnd(struct Node* head, int data) {
    struct Node* newNode = createNode(data);
    if (head == NULL)
        return newNode;

    struct Node* temp = head;
    while (temp->next != NULL)
        temp = temp->next;

    temp->next = newNode;
    newNode->prev = temp;
    return head;
}

struct Node* deleteNode(struct Node* head, int x) {
    if (head == NULL) return NULL;

    struct Node* temp = head;
    if (x == 1) {
        head = head->next;
        if (head != NULL)
            head->prev = NULL;
        free(temp);
        return head;
    }

    for (int i = 1; temp != NULL && i < x; i++)
        temp = temp->next;

    if (temp == NULL)
        return head;

    if (temp->next != NULL)
        temp->next->prev = temp->prev;

    if (temp->prev != NULL)
        temp->prev->next = temp->next;

    free(temp);
    return head;
}

void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d", head->data);
        if (head->next != NULL)
            printf(" <--> ");
        head = head->next;
    }
    printf("\n");
}

int main() {
    struct Node* head = NULL;

    head = insertEnd(head, 1);
    head = insertEnd(head, 5);
    head = insertEnd(head, 2);
    head = insertEnd(head, 9);

    int x = 1;
    head = deleteNode(head, x);

    printList(head);

    return 0;
}
